From: Harald Sitter <sitter@kde.org>
Date: Fri, 19 Dec 2014 15:47:40 +0000
Subject: prevent filewatches from emitting changes when there is no file
X-Git-Url: http://quickgit.kde.org/?p=qca.git&a=commitdiff&h=4aae2dee34d2f0f6324a9e7819e29310106dc5bb
---
prevent filewatches from emitting changes when there is no file

(this also fixes the flaky filewatch test that would fail for example on
ubuntu launchpad builds)

it can happen that (supposedly for filesystem reasons) there are two
changes signals arriving in the watcher code, by the time the first arrives
the file would however already be deleted, by the time the second arrives
it would thus notify of changes to a file that does not exist which is
silly and causes problems with the filewatchtest as it uses signal emission
counting to verify the behavior.
to prevent this problem from popping up there is an additional save guard
in the file_changed slot that will ignore a change if the path in question
doesn't exist AND there is no file being watched by the watcher.

REVIEW: 121588
---


--- a/src/support/dirwatch.cpp
+++ b/src/support/dirwatch.cpp
@@ -210,8 +210,19 @@
 	{
 		Q_UNUSED(path);
 		QFileInfo fi(filePath);
-		if(!fi.exists())
+		if (!fi.exists() && !fileExisted) {
+			// Got a file changed signal on a file that does not exist
+			// and is not actively watched. This happens when we
+			// previously watched a file but it was deleted and after
+			// the original deletion changed-signal we get another one
+			// (for example because of bad signal timing). In this scenario
+			// we must ignore the change as the change, whatever it may
+			// have been, is of no interest to us because we don't watch
+			// the file and furthermore the file does not even exist.
+			return;
+		} else if (!fi.exists()) {
 			fileExisted = false;
+		};
 		emit q->changed();
 	}
 };

